사용자 등록,삭제 서비스 테스트 코드 추가

youngjinmo opened the pull request:
사용자 등록과 삭제 API 테스트 코드를 추가했습니다!

youngjinmo committed:
사용자 등록,삭제 서비스 테스트 코드 추가

zaragit commented on src/test/java/com/bomstart/tobyspring/user/service/UserServiceTest.java:
```diff
@@ -0,0 +1,72 @@
+    UserService userService;
+
+    @AfterEach
+    @Test
+    @DisplayName("테스트 끝나고 테스트 유저 삭제")
```
해당 메소드는 @AfterEach로서 기능과 @Test로서 기능을 둘 다 가지고 있는거 같은데요.

아무래도 메소드 내에서 userService.deleteUser(userId)의 동작을 하다 보니 @AfterEach의 역할을 하면서 Test도 필요하다고 생각하고 작업하신 걸까요??

그렇다면 deleteUser 테스트 외로 추가적으로 deleteNull 같은 id가 null이 들어왔을 경우의 단위 테스트를 작성하고 @AfterEach에서는 Test와 null 체크를 제외하는 건 어떻게 생각하시는지 의견 듣고 싶습니다!!

youngjinmo replied:
> 해당 메소드는 @AfterEach로서 기능과 @test로서 기능을 둘 다 가지고 있는거 같은데요.

@AfterEach가 @Test가 중복되는지에 대해 이해가 부족했습니다.!

> 그렇다면 deleteUser 테스트 외로 추가적으로 deleteNull 같은 id가 null이 들어왔을 경우의 단위 테스트를 작성하고 @AfterEach에서는 Test와 null 체크를 제외하는 건 어떻게 생각하시는지 의견 듣고 싶습니다!!

동의합니다! 생각해보니 등록하고 삭제하는 과정에서도 여러가지 케이스가 있는데, 말씀하신 상황에 따른 테스트 코드를 작성하고, 테스트 코드를 바탕으로 실제 API도 수정을 해야할 것 같습니다.

zaragit commented on src/test/java/com/bomstart/tobyspring/user/service/UserServiceTest.java:
```diff
@@ -0,0 +1,72 @@
+        String expected = user.getName();
+        String actual = userService.getUser(user.getId()).getName();
+
+        System.out.println("expected : "+expected);
+        System.out.println("actual   : "+actual);
```
테스트 코드 작성이 끝난 이후로는 따로 로그가 필요하지 않을 것 같은데
혹시 이 부분은 작성 후에 지우시는 걸 깜빡하신 걸까요??

youngjinmo replied:
눈으로 확인하고 싶어서 남기긴 했습니다만, 사실 필요하지 않은 코드임에 동의합니다! 
이부분도 삭제하겠습니다!

zaragit reviewed (commented):
벌써 작업해 주셨네요. 제가 매번 늦네요... 😂
이번에도 먼저 너무 잘 작업해 주셔서 제가 작업할 때 굉장히 편할 것 같습니다. 

개인 적인 의견 2가지가 있어서 comment 남겼습니다! 의견 부탁 드립니다. 
코드는 Merge 해도 될 것 같습니다. 

저는 오후에 틈틈히 작업해서 저녁에 PR 하겠습니다.

youngjinmo commented:
피드백 감사합니다! 저도 수정하고 다시 PR하겠습니다! ㅎㅎ

youngjinmo committed:
fix. SQL 문법 수정
- update DaoImpl에서 SET 컬럼간 구분점(,) 추가

youngjinmo committed:
refactor. 코드 리뷰 피드백 반영
- @AfterEach와 @Test는 중복해서 사용할 필요없음
- 예외사항에 대해 조건문으로 제어하기 위한 로직을 컨트롤러에 추가
- 바뀐 로직을 테스트하기 위해 테스트 대상을 컨트롤러에서 서비스로 변경
- 삭제 요청 API에서 id가 null또는 존재하지 않는 회원일때 테스트

youngjinmo commented:
user update 관련 DaoImpl에 SQL 문법을 수정하였습니다!
SET 절에서 컬럼과 컬럼간 `,`가 빠져서 SQL예외가 발생하길래 `,`을 추가하였습니다.
이외 말씀하신 피드백 반영해보았습니다. 

기존에 코드가 심플하게 유저를 등록하고, 삭제만 가능했던것을 컨트롤러에서 유저를 검증하고 서비스 로직을 태우는것으로 변경하였습니다. 이에따라 테스트 코드 대상도 서비스에서 컨트롤러로 바꿔보았습니다.

천천히 확인해주시고 피드백 주시면 감사하겠습니다!!

zaragit commented on src/main/java/com/bomstart/tobyspring/user/Controller/UserController.java:
```diff
@@ -37,7 +44,12 @@ public UserController (UserService userService) {
     @DeleteMapping("/delete/{id}")
     public @ResponseBody void deleteUser(@PathVariable String id){
-        this.userService.deleteUser(id);
+        if(userService.getUser(id)!=null){
+            userService.deleteUser(id);
+            System.out.println("회원 정보를 삭제했습니다.");
+        } else {
+            System.out.println("해당 id를 가진 회원이 존재하지 않습니다.");
+        }
```
제 의견은 여기서 if 문은 비즈니스 로직이라고 생각이 됩니다. 그래서 Service 단에서 처리하는게 좋을 것 같습니다!!

MVC 패턴에서 Controller는 클라이언트로부터 받은 Request에 대해서 어떤 비즈니스 로직을 호출해 결과를 받아올지 결정하고, 받은 데이터를 어떻게 클라이언트에게 Response 할 지에만 집중하는 게 좋을 것 같습니다.

해당 내용 의견 부탁드립니다~

youngjinmo replied:
안그래도 이 부분에서 고민이 있었는데요, 병주님 말씀이 맞는것 같습니다. 
if문 또한 비즈니스 로직이므로 이는 서비스 로직으로 옮기고, 컨트롤러는 Request와 Response에만 집중하는게 MVC 패턴을 더 잘이용하는것 같습니다.!
감사합니다!

zaragit commented on src/test/java/com/bomstart/tobyspring/user/Controller/UserControllerTest.java:
```diff
@@ -0,0 +1,81 @@
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+@SpringBootTest
+class UserControllerTest {
+    @Autowired
```
테스트 대상을 Service에서 Controller 단으로 변경해 주셨네요!

바꾸게 된 이유가 Controller 단에 if 처리가 들어가면서 Controller부터 테스트를 진행해야 된다고 생각하셔서 변경해 주신 것 같은데 맞나요??

해당 부분 저도 궁금했던 내용이라서 Controller 테스트 코드는 어떻게 작성하나 찾아봤습니다.
이런 경우 MockMvc를 사용해서 실제 request 요청을 보내는 것처럼 코드를 작성하는 것 같습니다.

youngjinmo replied:
네네 if문이 들어가면서 컨트롤러를 대상으로 테스트를 했는데요, 위의 의견에 따라 테스트 대상은 서비스로 하는게 맞을것 같습니다!
Mock은 써본적이 없는데, 이 부분도 알아보겠습니다..!

zaragit reviewed (commented):
AfterEach에서 if 처리를 빼서 테스트 코드를 작성하는 것에 대해서 의견 드렸는데 벌써 작업까지 다 해주셨네요. 감사합니다!!
작업 과정에서 AfterEach를 완전히 제거해 주셨는데 혹시 기분 나쁘셨던 건 아닌지요... 😥

저도 영진님이 수정해 주신 코드 참고하면서 빠르게 코딩하고 있습니다.
또, 코딩 중에 의견이 생겨서 comment로 남겼습니다. 확인 부탁드립니다~

youngjinmo commented:
> 작업 과정에서 AfterEach를 완전히 제거해 주셨는데 혹시 기분 나쁘셨던 건 아닌지요... 😥

아닙니다..!ㅋㅋㅋ 작성해주신 피드백 모두 동의합니다!! A
fterEach의 경우는 지금처럼 쓰는게 최선이 아닌것 같아서 일단 지웠습니다..! 
개인적으로는 각 테스트가 끝나고서나서 SQL문의 Truncate를 사용해서 table에서 데이터를 제거하고 싶은데 이걸 AfterEach에 사용해도 괜찮을까요? 
이런 상황에서 AfterEach나 BeforeEach를 어떻게 사용하는게 좋을까요? 잘 모르겠어서 병주님의 의견을 여쭙고 싶습니다!

zaragit commented:
헛... 저도 테스트 실행 후에 테스터 데이터를 어떻게 해야할지 고민하고 있었습니다.

제가 생각하기엔 update, select에서도 작성해 주신
~~~java
        String expected = "TestUser";
        User user = new User();
        user.setId(expected);
        user.setName("TestName");
        user.setPassword("123456789");
~~~
테스트에 사용할 인스턴스를 만들고 있습니다.

deleteUser 호출 시 없는 데이터의 경우까지 테스트로 검증해 주셨으니
책에 나오는 것처럼 픽스처를 BeforeEach에서 만들고
만든 픽스처를 모두 AfterEach에서 deleteUser하는건 어떨까요??

찾아봤을 때 트랜잭션을 관리하는 어노테이션으로 간단하게 처리하는 경우도 있는 것 같아서 책 목차를 봤는데 
해당 내용은 이후 트랜잭션 챕터에 나오는 내용인 것 같습니다.

youngjinmo commented:
제가 책을 꼼꼼하게 안봤었군요..!
스크린샷속 코드처럼 하면 좋을것 같습니다! 감사합니다!!

youngjinmo committed:
refactor. 서비스 로직 리팩토링

youngjinmo committed:
test. 서비스 리팩토링 후 테스트 코드 작성

zaragit committed:
Merge branch 'main' into youngjin

zaragit committed:
Controller 테스트 코드 작성

youngjinmo committed:
Merge pull request #12 from zaragit/main

UserController 테스트 코드 작성

youngjinmo committed:
refactor. 불필요한 import 패키지 삭제

youngjinmo committed:
test. 사용자 등록/삭제 Service 테스트 코드 작성

youngjinmo committed:
test.  Mock을 이용한 API 테스트(회원 등록,삭제)

youngjinmo committed:
merge after conflict error

youngjinmo commented:
지난번에 요청해주신 Service 로직과 테스트를 수정하고, mock을 이용해서 Controller 테스트까지 진행했습니다!

Service와 Controller를 각각 통합 테스트를 하던 중에 회원목록 조회를 테스트하는 코드에서 expect보다 actual이 1이 더 높아서 테스트에 실패하는 케이스가 있었습니다. 단위 테스트를 진행하면 성공하는데 UserServiceTest를 통합테스트할때만 테스트에 실패하는데, 시간되실때 함께 찾아보면 좋을것 같습니다! 

제가 잠시 이것저것 코드를 만져보았지만, 원인을 아직까지 모르겠습니다..

이외 나머지 코드에 대해 리뷰해주시면 감사하겠습니다!!

zaragit commented:
Service, Controller 테스트 모두 정말 깔끔하게 잘 작성해 주신 것 같습니다!!
바로 Merge 해도 될 것 같습니다. 추가로 작업할 작업이 없으시면 바로 Merge 해도 될 것 같습니다.

회원 목록 조회 테스트에 문제가 있나보네요... ㅠㅠ 
제가 잘 확인 못한 부분이 있을 것 같습니다. 또, 코드 작성 중에 오류가 있었나 보네요...
퇴근하고 저녁에 바로 확인해 보겠습니다!!

youngjinmo commented:
@zaragit 금요일까지 시간두고 같이 찾아보면 좋을것 같습니다! 저도 시간나면 또 찾아보겠습니다!! 리뷰 감사합니다 병주님!
